src/IO/WriteBufferFromFileDecorator.cpp

```diff
@@ -15,10 +16,17 @@ void WriteBufferFromFileDecorator::finalizeImpl()
 {
     next();
 
+    /// In case of exception in preFinalize as a part of finalize call
+    /// WriteBufferFromFileDecorator.finalized is set as true
+    /// but impl->finalized is remain false
+    /// That leads to situation when the destructor of impl is called with impl->finalized equal false.
     if (!is_prefinalized)
         WriteBufferFromFileDecorator::preFinalize();
 
-    impl->finalize();
+    {
+        SwapHelper swap(*this, *impl);
+        impl->finalize();
```

Comment:
Can you explain what is going on here? Why do we need this `SwapHelper` and why can't we just call `impl->finalize()`?

Reply:
> @CheSema Can you explain what is going on here? Why do we need this `SwapHelper` and why can't we just call `impl->finalize()`?


First of all, is it clear why SwapHelper is used in nextImpl call?
If it is, then the answer is short: impl has to see its working_buffer, not the working_buffer(nullptr, nullptr) and pos(nullptr). Otherwise impl ends up in a situation when working_buffer and pos were driven away from him.

Long read:
 WriteBufferFromFileDecorator is created with nulls in working_buffer and pos. It is an empty buffer. At the constructor WriteBufferFromFileDecorator extract  working_buffer and pos from impl with swap call. Due to which WriteBufferFromFileDecorator::write writes data into working_buffer from impl. This is important part. When the working_buffer is filled out, WriteBufferFromFileDecorator::next is called. Here WriteBufferFromFileDecorator::bytes is updated (also important var) and WriteBufferFromFileDecorator::nextImpl called. In this part WriteBufferFromFileDecorator::nextImpl delegates the call to the impl->next. But Impl needs to see its working_buffer and pos, for it SwapHelper is called. Here the bug is hidden couse next is called right before.  That call provide working_buffer from impl to the impl in right way and impl is able to consume the data.

The same things happens when WriteBufferFromFileDecorator delegates any call to Impl. And it is a BUG that WriteBufferFromFileDecorator missed swaps in some functions including destructor.

Reply:
I don't understand why we do that swapping at all. I understood that the constructor of WriteBufferFromFileDecorator need to get `working_buffer` from `impl`. But why can't the constructor of `WriteBufferFromFileDecorator` just call `BufferBase::set()` and then in `nextImpl` set the position again? See for example how `ReadBufferWrapper` is implemented. Swapping is two-directional and it changes both sides and it's more difficult to understand. So I suppose if just `set()` can do the job it's better not use swapping.

Reply:
I wouldn't advocate swapping at all. Here I found the problem in swapping implementation and fixed it.
You suggestion how to implement delegation it in a different way seems good. We should try it to make things simpler. I do not with to introduce such changes right here, but I would be happy to try implement it in following PR.

Reply:
Maybe it's just lack of comments. But it's still unclear to me why do we need so many operations. Because the code

```
SwapHelper swap(*this, *impl);
impl->finalize();
```

means something like

```
tmp1 = this->working_buffer;
tmp2 = this->internal_buffer;
tmp3 = this->pos;
this->working_buffer = impl->working_buffer;
this->internal_buffer = impl->internal_buffer;
this->pos = impl->pos;
impl->working_buffer = tmp1;
impl->internal_buffer = tmp2;
impl->pos = tmp3;

try
{
    impl->finalize();

    tmp1 = this->working_buffer;
    tmp2 = this->internal_buffer;
    tmp3 = this->pos;
    this->working_buffer = impl->working_buffer;
    this->internal_buffer = impl->internal_buffer;
    this->pos = impl->pos;
    impl->working_buffer = tmp1;
    impl->internal_buffer = tmp2;
    impl->pos = tmp3;
}
catch (...)
{
    tmp1 = this->working_buffer;
    tmp2 = this->internal_buffer;
    tmp3 = this->pos;
    this->working_buffer = impl->working_buffer;
    this->internal_buffer = impl->internal_buffer;
    this->pos = impl->pos;
    impl->working_buffer = tmp1;
    impl->internal_buffer = tmp2;
    impl->pos = tmp3;

    throw;
}
```